### PR TITLE
policy: move PolicyMapState into policy package

### PR DESCRIPTION
**Pull request by ianvernon**
This is part of a broader effort to move most of the policy computation logic into the policy package itself, and out of endpoint. The idea is that in the future, we will have a single 'policy' object, which will be a field within `Endpoint`. The endpoint will consume this type, but will not be involved in its creation - that will be the responsibility of the policy package. This will make unit-testing policy generation easier, and will hide implementation details away from the endpoint package.

I don't like putting something related closely to the datapath implementation within the policy package, but in order to unify everything in one 'Policy' object for an endpoint, I think this is a necessary step to do so. Eventually, translation to `PolicyMapState` can be done via some translation into a datapath representation of the generated policy.

No functional changes are intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6275)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Review by nebril — APPROVED** [review]
LGTM aside from hound comments

**Review by aanm — DISMISSED** [review]
discussed offline

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Tests timed out : 
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8202

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Tests failed due to `JoinEP` in logs:

Build link: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8206/execution/node/96/log/

```2018-11-26T22:17:55.149900338Z level=warning msg="JoinEP: Failed to load program" containerID=4012c572fa datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=63020 error="Failed to replace Qdisc for lxcc0f6e5052bd9: Link not found" file-path=63020_next/bpf_lxc.o ipv4=10.10.0.68 ipv6="f00d::a0a:0:0:f62c" k8sPodName=/ subsys=datapath-loader veth=lxcc0f6e5052bd9```

**Comment by ianvernon** [comment]
test-me-please

**Review by joestringer — COMMENTED** [review]
This seems fine, just one thing it'd be nice to reduce the stuttering as the hound bot is suggesting.

**Comment by ianvernon** [comment]
@aanm I've duplicated the `PolicyKey` type into `pkg/policy` for now to avoid pulling in the `pkg/bpf` dependency while not factoring out `PolicyKey` into its own package. 

@joestringer will address Hound's comments after I get the OK from @aanm .

File: pkg/policy/policymapstate.go
Comment (tgraf) [reply pkg/policy/policymapstate.go]: This is legit

**Review by tgraf — DISMISSED** [review]
Looks awesome overall.

File: pkg/policy/policymapstate.go
Comment (ianvernon) [reply pkg/policy/policymapstate.go]: Fixed.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
@cilium/endpoint please review.

File: pkg/policy/mapstate.go
Comment (aanm) [inline pkg/policy/mapstate.go:58]: I think this structure needs a new name. This is storing proxy state AFAIK.

File: pkg/policy/mapstate.go
Comment (aanm) [inline pkg/policy/mapstate.go:73]: When can we remove this? cc @joestringer

File: pkg/policy/mapstate.go
Comment (aanm) [inline pkg/policy/mapstate.go:40]: Ok, we are getting somewhere. If this is supposed to be (B), then it should be in a different package. Here we are already mixing both A and C. There should be a package that can import both A and C so it can perform the translation from A to B and from B to C.

`[ (A) User Policy ] -> [ (B) Endpoint Policy ] -> [ (C) Datapath Policy ]`

The fields in this key structure are already datapath aware they have no relation with this policy package.

Comment (ianvernon) [reply pkg/policy/mapstate.go:40]: @aanm this is meant to be an iterative change. I agree that we need to get to B, but I want to do this in a piecemeal, easy-to-review way. The reason I am doing this is so I can encapsulate what currently is translated policy for the endpoint in a single type, which I am doing in 
https://github.com/cilium/cilium/pull/6270 (which is based on top of this PR). In order to undo the current way we do things in policy translation, I am of the belief that if we can decouple the endpoint from policy translation (which is what this PR seeks to do), then we can get to B. After #6270 is reviewed / merged, I can then work to decouple what the generated policy for an endpoint is from the datapath implementation. But, to do that, I need to do some of the translation bits within the policy package, which for now, will include this `MapState` type.

File: pkg/policy/mapstate.go
Comment (tgraf) [reply pkg/policy/mapstate.go:73]: Let's do it in a dedicated PR and not mix it into this cleanup

**Review by tgraf — APPROVED** [review]

**Pull request merged**
